patroni/config_generator.py

```diff
@@ -0,0 +1,274 @@
+    except psycopg.Error as e:
+        sys.exit(f'Failed to establish PostgreSQL connection: {e}')
+
+    with conn.cursor() as cur:
+        cur.execute("SELECT 1 FROM pg_roles WHERE rolname=%s AND rolsuper='t';", (su_params['username'],))
+        if cur.rowcount < 1:
+            sys.exit('The provided user does not have superuser privilege')
+
+        cur.execute("SELECT name, current_setting(name) FROM pg_settings \
+                     WHERE context <> 'internal' \
+                     AND source IN ('configuration file', 'command line', 'environment variable') \
+                     AND category <> 'Write-Ahead Log / Recovery Target' \
+                     AND setting <> '(disabled)' \
+                     OR name IN ('hba_file', 'ident_file', 'config_file', \
+                                 'data_directory', \
+                                 'listen_addresses', 'port', \
+                                 'max_connections', 'max_worker_processes', 'max_wal_senders', \
+                                 'max_replication_slots', 'max_locks_per_transaction', 'max_prepared_transactions', \
+                                 'hot_standby', \
+                                 'wal_level', 'wal_log_hints', \
+                                 'wal_keep_segments', 'wal_keep_size', \
+                                 'track_commit_timestamp');")
+
+        helper_dict = dict.fromkeys(['port', 'listen_addresses'])
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p == 'cluster_name':
+                config['scope'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command',
+                       'hba_file', 'ident_file', 'config_file'):
+                # write commands to the local config due to security implications
+                # write hba/ident/config_file to local config to ensure they are not removed later
+                config['postgresql'].setdefault('parameters', {})
+                config['postgresql']['parameters'][p] = v
+            elif p in ('port', 'listen_addresses'):
+                helper_dict[p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
+
+    conn.close()
+
+    connect_port = parsed_dsn.get('port', os.getenv('PGPORT', helper_dict['port']))
+    config['postgresql']['connect_address'] = f'{get_ip()}:{connect_port}'
+    config['postgresql']['listen'] = f'{helper_dict["listen_addresses"]}:{helper_dict["port"]}'
+
+    # it makes sense to define postgresql.pg_hba/pg_ident only if hba_file/ident_file are set to defaults
+    default_hba_path = os.path.join(config['postgresql']['data_dir'], 'pg_hba.conf')
+    if config['postgresql']['parameters']['hba_file'] == default_hba_path:
```

Comment:
I didn't check it myself, but wondering if postgres is always returning an absolute path.